### PR TITLE
Feat/devtooling 652 Integration Facebook resource

### PR DESCRIPTION
**Pull request by shrutisuryawanshigenesys**

### CODE REVIEW

File: genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+		return nil, resp, fmt.Errorf("Failed to get facebook integration request: %v %v", err, resp)
+	}
+	if facebookIntegrationRequests.Entities == nil || len(*facebookIntegrationRequests.Entities) == 0 {
+		return &allFacebookIntegrationRequests, resp, nil
+	}
+
+	allFacebookIntegrationRequests = append(allFacebookIntegrationRequests, *facebookIntegrationRequests.Entities...)
+
+	for pageNum := 2; pageNum <= *facebookIntegrationRequests.PageCount; pageNum++ {
+		const pageSize = 100
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go:115]: can keep this outside loop

File: genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			break
+		}
+
+		allFacebookIntegrationRequests = append(allFacebookIntegrationRequests, *facebookIntegrationRequests.Entities...)
+	}
+
+	return &allFacebookIntegrationRequests, resp, err
+}
+
+// getIntegrationFacebookIdByNameFn is an implementation of the function to get a Genesys Cloud integration facebook by name
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go:132]: can incorporate a caching solution for this !!? ( Look at Genesyscloud_user datasource)

File: genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+func getIntegrationFacebookIdByNameFn(ctx context.Context, p *integrationFacebookProxy, name string) (string, bool, *platformclientv2.APIResponse, error) {
+	facebookIntegrationRequests, resp, err := getAllIntegrationFacebookFn(ctx, p)
+	if err != nil {
+		return "", false, resp, err
+	}
+
+	if facebookIntegrationRequests == nil || len(*facebookIntegrationRequests) == 0 {
+		return "", true, resp, fmt.Errorf("No integration facebook found with name %s", name)
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go:140]: can distinguish this error with that on line 152. This error can suggest there are no facebookIntegrationRequests at all.

File: genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			return retry.NonRetryableError(util.BuildWithRetriesApiDiagnosticError(resourceName, fmt.Sprintf("Failed to read integration facebook %s: %s", d.Id(), getErr), resp))
+		}
+
+		resourcedata.SetNillableValue(d, "name", facebookIntegrationRequest.Name)
+		resourcedata.SetNillableValue(d, "supported_content_id", facebookIntegrationRequest.SupportedContent.Id)
+		resourcedata.SetNillableValue(d, "messaging_setting_id", facebookIntegrationRequest.MessagingSetting.Id)
+		resourcedata.SetNillableValue(d, "page_id", facebookIntegrationRequest.PageId)
+		resourcedata.SetNillableValue(d, "app_id", facebookIntegrationRequest.AppId)
+
+		log.Printf("Read integration facebook")
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook.go:87]: add id here for better context, else remove the log if deemed unnecessary.

File: genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_utils.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	pageAccessToken := d.Get("page_access_token").(string)
+	userAccessToken := d.Get("user_access_token").(string)
+	pageId := d.Get("page_id").(string)
+	appId := d.Get("app_id").(string)
+	appSecret := d.Get("app_secret").(string)
+
+	if pageAccessToken != "" {
+		pageId = ""
+		userAccessToken = ""
+	} else {
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_utils.go:29]: else can be removed ?

File: genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_utils.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+	supportedContentId := d.Get("supported_content_id").(string)
+	messagingContentId := d.Get("messaging_setting_id").(string)
+	pageAccessToken := d.Get("page_access_token").(string)
+	userAccessToken := d.Get("user_access_token").(string)
+	pageId := d.Get("page_id").(string)
+	appId := d.Get("app_id").(string)
+	appSecret := d.Get("app_secret").(string)
+
+	if pageAccessToken != "" {
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_utils.go:26]: does this mean pageAccessToken is given precedence over useraccestoken ?

File: genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_schema.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+				Type:        schema.TypeString,
+				ForceNew:    true,
+			},
+			`app_id`: {
+				Description: `The app Id of Facebook app. The appId is required when a customer wants to use their own approved Facebook app.`,
+				Optional:    true,
+				Type:        schema.TypeString,
+				ForceNew:    true,
+			},
+			`app_secret`: {
```
Comment (HemanthDogiparthi12) [inline genesyscloud/integration_facebook/resource_genesyscloud_integration_facebook_schema.go:90]: may be better to add a validation inside the createContext just to check all these conditions are met. (ideal place would be a validateFunc but there are challenges referencing other input varaible in a current variable context)

if appsecret has a value entered, appId should be present too. (viceversa)
pageaccesstoken and userAccessToken need not be simultaneously present 
pageid need not be provided if pageaccesstoken is provided and userAccessToken not provided

File: docs/resources/integration_facebook.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+* [PATCH /api/v2/conversations/messaging/integrations/facebook/{integrationId}](https://developer.genesys.cloud/devapps/api-explorer#patch-api-v2-conversations-messaging-integrations-facebook--integrationId-)
+* [DELETE /api/v2/conversations/messaging/integrations/facebook/{integrationId}](https://developer.genesys.cloud/devapps/api-explorer#delete-api-v2-conversations-messaging-integrations-facebook--integrationId-)
+* [GET /api/v2/conversations/messaging/integrations/facebook](https://developer.genesys.cloud/devapps/api-explorer#get-api-v2-conversations-messaging-integrations-facebook)
+
+
+
+## Example Usage
+
+```terraform
+resource "genesyscloud_integration_facebook" "test_sample" {
```
Comment (HemanthDogiparthi12) [inline docs/resources/integration_facebook.md:25]: can you add one more example with user access token and pageid

File: genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	var facebookIntegration platformclientv2.Facebookintegration
+	for _, facebookIntegrationRequest := range *facebookIntegrationRequests {
+		if *facebookIntegrationRequest.Name == name {
+			log.Printf("Retrieved the integration facebook id %s by name %s", *facebookIntegrationRequest.Id, name)
+			facebookIntegration = facebookIntegrationRequest
+			return *facebookIntegration.Id, false, resp, nil
+		}
+	}
+
+	return "", true, resp, err
```
Comment (charliecon) [inline genesyscloud/integration_facebook/genesyscloud_integration_facebook_proxy.go:160]: Here you will be returning `nil` for the error when you should be return `fmt.Errorf("No integration facebook found...")` as above

**Review by charliecon — COMMENTED** [review]
For me, the test `TestAccResourceIntegrationFacebook` is failing with the error message 

```
        Error: Failed to update integration facebook: API Error: 400 - Bad request: Create integration has not completed 

```

And the data source test is failing with the following: 
```
 Error: No integration facebook found with name test_sample
```

It looks like we might need a sleep after the create, or the maybe the retry logic in the read needs to be changed a bit - longer wait between retries, more retries, etc.

**Review by charliecon — APPROVED** [review]

**Pull request merged**